### PR TITLE
Fixed Quote Policy with Royalty

### PR DESCRIPTION
**Pull request by ggobugi27**
A version of fixed-quote-policy with royalty paid to creator. 
- `fungible`, `creator`, `creator-guard`, `royalty-rate` is saved to token via `enforce-init`
- `creator` must be valid account in `fungible`
- `royalty-rate` must be `>= 0.0` and `<= 1.0`
-  royalty is paid to the creator in `fungible` at the second step of `sale` via `enforce-buy`

### CODE REVIEW

File: pact/fixed-quote-policy-royalty.pact [hunk 1]
```diff
@@ -0,0 +1,192 @@
+(namespace (read-msg 'ns))
+
+(module fixed-quote-policy-royalty GOVERNANCE
```
Comment (sirlensalot) [inline pact/fixed-quote-policy-royalty.pact:3]:
```suggestion
(module fixed-quote-royalty-policy GOVERNANCE
```

File: pact/fixed-quote-policy-royalty.pact [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+  @doc "Policy for fixed issuance with simple quoted sale."
```
Comment (sirlensalot) [inline pact/fixed-quote-policy-royalty.pact:5]:
```suggestion
  @doc "Policy for fixed issuance with royalty and quoted sale in specified fungible."
```

File: pact/fixed-quote-policy-royalty.pact [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    (enforce-guard (keyset-ref-guard 'marmalade-ns-admin )))
+
+  (implements kip.token-policy-v1_DRAFT3)
+  (use kip.token-policy-v1_DRAFT3 [token-info])
+
+  (defschema policy-schema
+    fungible:module{fungible-v2}
+    creator:string
+    creator-guard:guard
+    mint-guard:guard
```
Comment (sirlensalot) [inline pact/fixed-quote-policy-royalty.pact:17]: I think `creator-guard` can also be `mint-guard`, no?

File: pact/fixed-quote-policy-royalty.pact [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    )
+    (enforce-ledger)
+    (enforce false "Burn prohibited")
+  )
+
+  (defun enforce-init:bool
+    ( token:object{token-info}
+    )
+    (enforce-ledger)
+    (let* ( (fungible:module{fungible-v2} (read-msg 'fungible ))
```
Comment (sirlensalot) [inline pact/fixed-quote-policy-royalty.pact:76]: Hmm, shouldn't we just be using `policy-schema` in the payload here? smth like `(read-msg TOKEN_SPEC)` and then bind on `fungible` `creator` etc

File: pact/fixed-quote-policy-royalty.pact [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+            (royalty-rate:decimal (read-decimal 'royalty-rate ))
+            (creator-details:object (fungible::details creator ))
+            )
+      (fungible::precision)
+      (enforce (=
+        (at 'guard creator-details) creator-guard)
+        "Creator guard does not match")
+      (enforce (and
+        (>= royalty-rate 0.0) (<= royalty-rate 1.0))
+        "royalty rate is not valid")
```
Comment (sirlensalot) [inline pact/fixed-quote-policy-royalty.pact:91]:
```suggestion
        "Invalid royalty rate")
```
(avoid "is not" in error messages when possible, it's shorter and less conversational)

File: pact/fixed-quote-policy-royalty.pact [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            (price:decimal (at 'price spec))
+            (recipient:string (at 'recipient spec))
+            (recipient-guard:guard (at 'recipient-guard spec))
+            (recipient-details:object (fungible::details recipient)) )
+      (fungible::enforce-unit price)
+      (enforce (< 0.0 price) "Offer amount must be positive")
+      (enforce (=
+        (at 'guard recipient-details) recipient-guard)
+        "Recipient guard does not match")
+      (insert quotes sale-id { 'id: (at 'id token), 'spec: spec }))
```
Comment (sirlensalot) [inline pact/fixed-quote-policy-royalty.pact:125]: We need to enforce that the royalty rate doesn't run into precision here. Ie a royalty rate of 1% with a fungible precision of 4 and a price 0.0001 would result in an invalid royalty (so the seller must increase the quote price).

File: pact/fixed-quote-policy-royalty.pact [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+      , 'royalty-rate:= royalty-rate:decimal
+      }
+      (with-read quotes sale-id { 'id:= qtoken, 'spec:= spec:object{quote-spec} }
+        (enforce (= qtoken (at 'id token)) "incorrect sale token")
+        (bind spec
+          { 'price := price:decimal
+          , 'recipient := recipient:string
+          , 'recipient-guard := recipient-guard:guard
+          }
+          (fungible::transfer-create buyer creator creator-guard (* (* amount price) royalty-rate))
```
Comment (sirlensalot) [inline pact/fixed-quote-policy-royalty.pact:149]: let bind `(sale-price (* amount price))` for use here and next line

File: pact/fixed-quote-policy-royalty.pact (hunk 3 above)
Comment (ggobugi27) [reply pact/fixed-quote-policy-royalty.pact]: `creator-guard` is also used for the account guard in the designated fungible that will receive the fund, so I think `mint-guard` can serve it's own purpose? If the creator wants to use the same guard, they can put the same key in the guard, but they might not want to unify them .

File: pact/fixed-quote-policy-royalty.pact (hunk 6 above)
Comment (ggobugi27) [reply pact/fixed-quote-policy-royalty.pact]: Enforced `(* sale-price royalty-rate)`. Also added `(enforce-unit royalty-rate)` in `enforce-init`

File: pact/marmalade.repl [hunk 8]
```diff
@@ -1041,6 +1041,197 @@
...
+
+(expect "coinbase buyer" "Write succeeded"
+  (coin.coinbase "coin-buyer" (read-msg 'buyer-guard) 2.0))
+
+(expect "coin-buyer is credited coins"
+  2.0
+  (coin.get-balance 'coin-buyer))
+
+(expect-failure
+  "0.00000000000006"
```
Comment (sirlensalot) [inline pact/marmalade.repl:1149]: This is missing a description maybe?

File: pact/marmalade.repl [hunk 9]
```diff
@@ -1041,6 +1041,197 @@
...
+  ,"buyer-guard": {"keys": ["buyer"], "pred": "="}
+  })
+(env-chain-data {"block-height": 100})
+
+(test-capability (coin.COINBASE))
+
+(expect "coinbase buyer" "Write succeeded"
+  (coin.coinbase "coin-buyer" (read-msg 'buyer-guard) 2.0))
+
+(expect "coin-buyer is credited coins"
```
Comment (sirlensalot) [inline pact/marmalade.repl:1144]: don't need this test

File: pact/marmalade.repl [hunk 10]
```diff
@@ -1041,6 +1041,197 @@
...
+    (map (remove 'module-hash) (env-events true)))
+
+(commit-tx)
+
+(begin-tx "sale fixed-quote-royalty-policy success")
+(env-hash (hash "fixed-quote-royalty-policy-sale-tx2"))
+(use marmalade.ledger)
+
+(env-data {"alice": {"keys": ["alice"], "pred":"="}})
+(expect
```
Comment (sirlensalot) [inline pact/marmalade.repl:1116]: generally for fixtures/other modules don't use `expect` etc as it is not properly a use case for this test. Just call `coin.create-account ...` directly

File: pact/marmalade.repl [hunk 11]
```diff
@@ -1041,6 +1041,197 @@
...
+   ,'caps: [(marmalade.ledger.BUY "project-3" "bob" "coin-buyer" 0.2 120 (hash "fixed-quote-royalty-policy-sale-tx2"))]
+  }
+,{'key: 'dummy
+ ,'caps: [
+   (coin.TRANSFER "coin-buyer" "alice" 0.194)
+   (coin.TRANSFER "coin-buyer" "creator" 0.006)
+ ]}])
+
+(expect-failure "buy fails, buyer doesn't sign coin.TRANSFER"
+  "[buyer]"
```
Comment (sirlensalot) [inline pact/marmalade.repl:1197]: hmm that's weird error text, what's going on there?

File: pact/marmalade.repl [hunk 12]
```diff
@@ -1041,6 +1041,197 @@
...
+  {'key: 'bob
+   ,'caps: [(marmalade.ledger.BUY "project-3" "bob" "coin-buyer" 0.2 120 (hash "fixed-quote-royalty-policy-sale-tx2"))]
+  }
+ ,{'key: 'buyer
+ ,'caps: [
+  (coin.TRANSFER "coin-buyer" "alice" 0.194)
+  (coin.TRANSFER "coin-buyer" "creator" 0.006)
+ ]}])
+
+(expect "buy suceeds"
```
Comment (sirlensalot) [inline pact/marmalade.repl:1210]:
```suggestion
(expect "buy succeeds"
```

File: pact/marmalade.repl [hunk 13]
```diff
@@ -1041,6 +1041,197 @@
...
+  true
+  (continue-pact 1))
+
+(expect "Seller is credited, sale account is debited on project-3 balance"
+  [0.8 0.0 0.2]
+  (map (get-balance 'project-3) ['bob (sale-account) 'coin-buyer]))
+
+(expect "coin-buyer is debited on coin balance by price"
+  [1.8 0.194 0.006]
```
Comment (sirlensalot) [inline pact/marmalade.repl:1219]: Document the values here aka 
```[ 1.8 ;; buyer debited for sale price 2.0
, 0.194 ;; seller credited sale price minus 3% royalty
, 0.006 ;; creator 3% royalty received
```

although it's probably clearer just to have 3 `expect` calls with appropriate descriptions.

**Review by sirlensalot — COMMENTED** [review]
Looks great! Test comments

File: pact/marmalade.repl (hunk 11 above)
Comment (ggobugi27) [reply pact/marmalade.repl:1197]: Error text is from guard failure at `coin.TRANSFER`, where buyer keyset is `{"keys": ["buyer"], "pred": "="}`

**Review by sirlensalot — APPROVED** [review]

**Pull request merged**